### PR TITLE
[CMSIS-NN] Aligned buffer size APIs to match with CMSIS-NN

### PR DESCRIPTION
**Pull request by ashutosh-arm**
This commit introduces new data structures that
closely match with CMSIS-NN data structures.
New buffer size APIs were derived from CMSIS-NN
equivalents that make use of these new objects.
This will help with automation of buffer size
API generation while updating CMSIS-NN version
in TVM in the future.

### CODE REVIEW

**Review by areusch — COMMENTED** [review]
thanks @ashutosh-arm , asked a couple questions here

File: src/relay/backend/contrib/cmsisnn/buffer_size.cc [hunk 1]
```diff
@@ -16,141 +16,198 @@
...
-    size = Conv2dBufferSizeInt16(target, padding_w, padding_h, input_n, input_h, input_c, output_h,
-                                 output_w, stride_w, stride_h, dilation_w, dilation_h, filter_w,
-                                 filter_h);
+#define USE_FAST_DW_CONV_FUNCTION(dw_conv_params, filter_dims, input_dims) \
+  dw_conv_params->ch_mult == 1 && dw_conv_params->dilation.w == 1 &&       \
+      dw_conv_params->dilation.h == 1 && filter_dims->w * filter_dims->h * input_dims->c < 512
+
+#define CH_IN_BLOCK_MVE (124)
+
+int32_t AvgpoolS16GetBufferSize(bool has_mve, bool has_dsp, const int output_x, const int ch_src) {
```
[the remark this answers is not in the document]
Comment (ashutosh-arm) [reply src/relay/backend/contrib/cmsisnn/buffer_size.cc:41]: This file is derived from one of the files in CMSIS-NN headers. There is a link to the appropriate file from which these APIs were taken keeping in mind no direct duplication. Do we still need to add comments?

File: src/relay/backend/contrib/cmsisnn/types.h [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file src/relay/backend/contrib/cmsisnn/types.h
+ * \brief Includes structs and data types needed by CMSIS-NN APIs.
```
[the remark this answers is not in the document]
Comment (ashutosh-arm) [reply src/relay/backend/contrib/cmsisnn/types.h:22]: @u99127 can better comment about it.

File: src/relay/backend/contrib/cmsisnn/buffer_size.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (u99127) [reply src/relay/backend/contrib/cmsisnn/buffer_size.cc:41]: We just discussed this offline - one of the things to investigate is whether there could be a script to auto-generate this file without having to do this manually. If so, we could investigate putting such a script into the tvm codebase.

File: src/relay/backend/contrib/cmsisnn/types.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (u99127) [reply src/relay/backend/contrib/cmsisnn/types.h:22]: I think there's more work here to put CMSIS-NN source into 3rd party especially as this is a static library which will need different builds for different micro-architectures for the runtime aspects . Can we take that up as a separate activity for the future ? 

I would also want to circle back on this one.

**Comment by ashutosh-arm** [comment]
Closing this PR as we expect more support for API auto-generation into CMSIS-NN: https://github.com/ARM-software/CMSIS-NN. @Mousius

**Pull request closed without merging**